File: src/main/java/io/github/chaosdave34/kitpvp/kits/ElytraKit.java
package io.github.chaosdave34.kitpvp.kits;

import com.mojang.datafixers.util.Pair;
import io.github.chaosdave34.kitpvp.ExtendedPlayer;
import lombok.Getter;
import org.bukkit.Material;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.bukkit.inventory.meta.FireworkMeta;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

@Getter
public abstract class ElytraKit extends Kit {
    protected Material icon;

     public ElytraKit(String id, String name, Material icon) {
         super(id, name);
         this.icon = icon;
     }

    @Override
    public ItemStack getChestContent() {
        return new ItemStack(Material.ELYTRA);
    }

    @Override
    public ItemStack getOffhandContent() {
         ItemStack firework = new ItemStack(Material.FIREWORK_ROCKET, 3);
         firework.editMeta(FireworkMeta.class, fireworkMeta -> fireworkMeta.setPower(2));

        return firework;
    }

    @Override
    public void apply(Player p) {
        ExtendedPlayer extendedPlayer = ExtendedPlayer.from(p);
        extendedPlayer.setSelectedElytraKitId(id);

        AttributeInstance maxHealth = p.getAttribute(Attribute.GENERIC_MAX_HEALTH);
        if (maxHealth != null) maxHealth.setBaseValue(getMaxHealth());
        p.setHealth(getMaxHealth());

        AttributeInstance movementSpeed = p.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED);
        if (movementSpeed != null) movementSpeed.setBaseValue(getMovementSpeed());

        PlayerInventory inv = p.getInventory();

        inv.setContents(getInventoryContent());
        inv.setHelmet(getHeadContent());
        inv.setChestplate(getChestContent());
        inv.setLeggings(getLegsContent());
        inv.setBoots(getFeetContent());
        inv.setItemInOffHand(getOffhandContent());
        inv.addItem(getKillRewards());

        inv.setItem(17, new ItemStack(Material.ARROW));

        for (PotionEffect effect : p.getActivePotionEffects()) {
            p.removePotionEffect(effect.getType());
        }

        for (Pair<PotionEffectType, Integer> potionEffect : getPotionEffects()) {
            p.addPotionEffect(new PotionEffect(potionEffect.getFirst(), -1, potionEffect.getSecond(), false, false, false));
        }

        extendedPlayer.removeCompanion();
        extendedPlayer.spawnCompanion();
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/kits/ElytraKit.kt
package io.github.chaosdave34.kitpvp.kits

import io.github.chaosdave34.kitpvp.ExtendedPlayer
import org.bukkit.Material
import org.bukkit.attribute.Attribute
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.FireworkMeta
import org.bukkit.potion.PotionEffect

abstract class ElytraKit(id: String, name: String, val icon: Material) : Kit(id, name) {

    override fun getChestContent(): ItemStack = ItemStack.of(Material.ELYTRA)

    override fun getOffhandContent(): ItemStack {
        val firework = ItemStack.of(Material.FIREWORK_ROCKET, 3)
        firework.editMeta(FireworkMeta::class.java) { fireworkMeta -> fireworkMeta.power = 3 }
        return firework
    }

    override fun apply(player: Player) {
        val extendedPlayer = ExtendedPlayer.from(player)
        extendedPlayer.setSelectedElytraKitId(id)

        player.getAttribute(Attribute.GENERIC_MAX_HEALTH)?.baseValue = getMaxHealth()
        player.health = getMaxHealth()

        player.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED)?.baseValue = getMovementSpeed()

        val inventory = player.inventory

        inventory.contents = getInventoryContent()
        inventory.helmet = getHeadContent()
        inventory.chestplate = getChestContent()
        inventory.leggings = getLegsContent()
        inventory.boots = getFeetContent()
        inventory.setItemInOffHand(getOffhandContent())
        inventory.addItem(*getKillRewards())

        inventory.setItem(17, ItemStack.of(Material.ARROW))

        player.clearActivePotionEffects()

        for ((effectType, amplifier) in getPotionEffects()) {
            player.addPotionEffect(PotionEffect(effectType, -1, amplifier, false, false, false))
        }

        extendedPlayer.removeCompanion()
    }

    override fun Player.isKitSelected(): Boolean = id == ExtendedPlayer.from(this).selectedElytraKitId

    override fun Player.isKitActive(): Boolean = isKitSelected() && ExtendedPlayer.from(this).gameState == ExtendedPlayer.GameState.ELYTRA_IN_GAME
}